### PR TITLE
Merge agent skill and jupyter

### PR DESCRIPTION
**Pull request by Shimada666**
#2016 
@xingyaoww 
After integrating the agent skill library, the agent_skill itself depends on Jupyter. Currently, both plugins are almost always imported simultaneously. Therefore, we can merge agent_skill and Jupyter.

### CODE REVIEW

**Comment by enyst** [comment]
I don't know about this, I wouldn't do it I think... Jupyter is essential, skills are changing. There can be mandatory skills and optional skills at some point. As it is now, perhaps we could write a second file to provide to Jupyter on our own machine and use as template the existing one? Perhaps with editing less files.

I'm not sure it's better to make it into a more monolithic approach. Please note that I don't know this setup well, it might be that I'm missing something. Does the monolithic approach have significant performance improvement?

File: agenthub/codeact_agent/codeact_agent.py [hunk 1]
```diff
@@ -148,7 +147,6 @@ class CodeActAgent(Agent):
         # AgentSkillsRequirement provides a lot of Python functions
```
Comment (isavita) [inline agenthub/codeact_agent/codeact_agent.py:147]: Probably this NOTE: is not needed anymore.

File: opendevin/runtime/plugins/agent_skills/setup.sh [hunk 2]
```diff
@@ -11,3 +15,72 @@ echo 'export PYTHONPATH=/opendevin/plugins/agent_skills:$PYTHONPATH' >> ~/.bashr
...
 pip install flake8 python-docx PyPDF2 python-pptx pylatexenc openai opencv-python
+
+# ======================
+# Jupyter
+# ======================
+source ~/.bashrc
+# if user name is `opendevin`, add '/home/opendevin/.local/bin' to PATH
+if [ "$USER" = "opendevin" ]; then
+    echo 'export PATH=$PATH:/home/opendevin/.local/bin' >> ~/.bashrc
+    export PATH=$PATH:/home/opendevin/.local/bin
```
Comment (yufansong) [inline opendevin/runtime/plugins/agent_skills/setup.sh:26]: Do we need to it twice? I guess maybe we can echo them into .bashrc then source it in the end.

**Comment by Shimada666** [comment]
@enyst 

Thank you for your review, and I understand your concerns! You are probably worried that having too many skills might lead to bloating with numerous plugins, resulting in longer dependency installation times and larger sandbox images.

You don't need to worry about that!😉 In the #2016 discussion, we mentioned future architectural improvements. Subsequently, functions like `parse_audio` will be transformed into an `AudioToTextAction`, which will directly send requests to the backend REST API to obtain responses. The actual logic won't run in the sandbox anymore (e.g., calling openai).

Architecture: SANDBOX Jupyter <---> Rest API <---> LLM. After the architectural improvements, we won't need to run real logic in the sandbox, and naturally, we won't have to install too many dependencies.

I believe that merging the existing agent skill with Jupyter directly should not be a problem. The existing agent skills can serve as the foundational multimodal capability. 😁 

> I discussed with @rbren earlier, I think it will a more fundamental solution if we can connect these agentskills (executed inside sandbox) with Event Stream on our backend (e.g., every time a audio-to-text function is called by the agent inside sandbox, it push an AudioToTextAction to the EventStream on the backend - maybe by exposing a REST API from backend; The python function execution inside sandbox will be blocked until an AudioToTextObservation is processed by the backend and returned to the function in the sandbox).
Originally posted by @xingyaoww in https://github.com/OpenDevin/OpenDevin/pull/2016#discussion_r1614280552

cc @xingyaoww

File: opendevin/runtime/plugins/agent_skills/setup.sh (hunk 2 above)
Comment (Shimada666) [reply opendevin/runtime/plugins/agent_skills/setup.sh]: Sure! updated.

File: agenthub/codeact_agent/codeact_agent.py (hunk 1 above)
Comment (Shimada666) [reply agenthub/codeact_agent/codeact_agent.py]: Thanks! removed.

**Comment by Shimada666** [comment]
@enyst @xingyaoww 
If we are planning to expand with a significant number of plugins in the future, making the plugins optional is indeed necessary.

Currently, there is an issue where `AgentSkillsRequirement` must be loaded before `JupyterRequirement`. If the import order of Requirements is incorrect in `PluginsRequirement.init_plugins`, it will not function properly. My idea is to add a depends_on option in PluginRequirement, for example, `AgentSkillsRequirement` depends_on [`JupyterRequirement`]. Simultaneously, rewriting the logic in `init_plugins` to automatically load based on priority will eliminate the need for enforcing import order.

At the same time, in our sandbox Dockerfile, we only install basic dependencies (such as `Jupyter`). Additional dependencies are still installed in the `setup.sh` of the plugins. If needed in the future, we can consider customizing a sandbox docker image for each agent, containing all the dependencies required by the agent.

However, as it stands, `AgentSkillsRequirement` and `JupyterRequirement` are basically needed together... They serve as basic plugins and are still very appropriate, so I have submitted this PR. I am looking forward to hearing your thoughts on this!!

**Pull request closed without merging**